### PR TITLE
DS-2187 : Remove Lucene search support

### PR DESCRIPTION
**Pull request by tdonohue**
https://jira.duraspace.org/browse/DS-2187

This PR completely removes all APIs, configurations and views (JSPs / XMLUI aspects) which were associated with Lucene. It has been lightly tested both on XMLUI and JSPUI, and everything seems to work, but it could use additional eyes/testers.

This work is based off of initial work from @kshepherd. I borrowed his [DS-2187 branch](https://github.com/kshepherd/DSpace/tree/DS-2187) and completed the work by refactoring it against the latest master (i.e. service api) and removing additional classes/configurations.

I've tried to do as thorough a code / configuration removal as I could. However, it's always possible I've overlooked something that is no longer needed (especially i18n messages).

This is ready for others to test/analyze for possible inclusion in 6.0.

### CODE REVIEW

File: dspace-xmlui/src/main/webapp/sitemap.xmap [hunk 1]
```diff
@@ -542,8 +541,12 @@
 
                     <!-- OpenSearch results document for whole repo or community/collection -->
                     <map:match pattern="open-search/**">
+                       <!-- Prior to DSpace 6.0, "open-search/" was mapped to the StandardOpenSearchGenerator,
+                            which was removed as part of removal of the legacy Lucene search index classes.
+                            This pattern has been removed to DiscoveryOpenSearchGenerator to help ensure
+                            minimal impact from removal of the legacy Lucene search index. -->
                         <map:match pattern="open-search/">
-                            <map:generate type="StandardOpenSearchGenerator"/>
+                            <map:generate type="DiscoveryOpenSearchGenerator"/>
```
Comment (tdonohue) [inline dspace-xmlui/src/main/webapp/sitemap.xmap]: NOTE: this PR also fixes [DS-2268](https://jira.duraspace.org/browse/DS-2268) as the old OpenSearch generator is now replaced by the Discovery driven one.

**Comment by tdonohue** [comment]
Sidenote: For some reason the changes in PR #929 are now missing from `master` (must have gotten lost during the Service API refactor). So, this PR also re-applies #929 (see https://github.com/tdonohue/DSpace/commit/501e40d87ef00dae7d1f1b6ea22e70be60ef35d4) as it's necessary to fully remove all references to Lucene APIs.

**Comment by hardyoyo** [comment]
Tested in Vagrant-DSpace, compiles cleanly and deploys, all webapps start (i.e. passes smoke test), I've clicked around a bit and haven't observed any glaring problems in the dspace.log file. I'll test a bit more over the weekend, so far this looks just fine.

**Comment by hardyoyo** [comment]
Tested a bit more today, I'm trying to confirm that this PR also addresses DS-2268 as tdonohue mentions above. I believe we may need to change some default configuration in order to properly say DS-2268 is addressed. I *think* Andrea Schweer's comment on 2268 from October 15, 2015 outlines the changes required. However, whenever I attempt to add my Vagrant-DSpace site as a search provider in Firefox, I get the following error from Firefox: "Firefox could not download the search plugin from:
http://dspace.vagrant.dev:8080/xmlui/open-search/description.xml"

**Comment by aschweer** [comment]
@hardyoyo, can you download the description.xml file directly (without going via the search provider route -- eg go to that URL with Firefox or use curl/wget)? What are the contents of that file? I have vague memories that I saw & fixed this at some point but I don't remember the details. Hopefully, seeing what's in that file will jog my memory.

**Comment by aschweer** [comment]
@hardyoyo oh also, you did enable opensearch in dspace.cfg before you tried this, correct? Default is disabled:
https://github.com/DSpace/DSpace/blob/master/dspace/config/dspace.cfg#L1395
`websvc.opensearch.enable = false`

**Comment by KevinVdV** [comment]
Haven't tested the pull request yet but took a scan at the code, so if it works +1 from me.

**Comment by hardyoyo** [comment]
@aschweer yep, I enabled opensearch as per the docs. I'll see if I can get you a copy of the description.xml file (I did download it directly).

**Comment by hardyoyo** [comment]
I've made a discovery (pardon the accidental pun), the configuration of the wbsvc.opensearch.svccontext value requires an ending slash.

Here is what I have set in my dspace.cfg:

websvc.opensearch.enable = true
websvc.opensearch.uicontext = discover
websvc.opensearch.svccontext = discover/
websvc.opensearch.autolink = true

This seems to work to add a new search provider to Firefox. Looking at the description.xml I can see the host name has been set to localhost, instead of the configured host name (dspace.vagrant.dev). In my particular test case, localhost works, but that won't always be the case. Here's the contents of description.xml:

<?xml version="1.0" encoding="UTF-8"?>
<OpenSearchDescription xmlns="http://a9.com/-/spec/opensearch/1.1/">
<ShortName>DSpace</ShortName>
<LongName>DSpace on Vagrant</LongName>
<Description>DSpace on Vagrant DSpace repository</Description>
<InputEncoding>UTF-8</InputEncoding>
<OutputEncoding>UTF-8</OutputEncoding>
<Query role="example" searchTerms="photosynthesis"/>
<Tags>IR DSpace</Tags>
<Contact>dspace@myu.edu</Contact>
<Image height="16" type="image/vnd.microsoft.icon" width="16">http://www.dspace.org/images/favicon.ico</Image>
<Url template="http://localhost:8080/xmlui/discover?query={searchTerms}" type="text/html"/>
<Url template="http://localhost:8080/xmlui/open-search/discover/?query={searchTerms}&amp;start={startIndex?}&amp;rpp={count?}&amp;format=atom" type="application/atom+xml; charset=UTF-8"/>
<Url template="http://localhost:8080/xmlui/open-search/discover/?query={searchTerms}&amp;start={startIndex?}&amp;rpp={count?}&amp;format=rss" type="application/rss+xml; charset=UTF-8"/>
</OpenSearchDescription>

**Comment by hardyoyo** [comment]
[description.xml.txt](https://github.com/DSpace/DSpace/files/88003/description.xml.txt)

**Comment by hardyoyo** [comment]
Setting down the distraction of DS-2268, I also noticed [index-db-browse](https://github.com/tdonohue/DSpace/blob/ce86b9c512b54d1e1afb5241438643400470eb0f/dspace/config/launcher.xml#L127-L133) remains in the launcher, as does the corresponding dspace-src/dspace-api/src/main/java/org/dspace/browse/IndexBrowse.java file.

**Comment by aschweer** [comment]
* It doesn't do the localhost thing for me (see eg http://researchcommons.waikato.ac.nz/open-search/description.xml) -- looks like this is generated [here](https://github.com/tdonohue/DSpace/blob/DS-2187-remove-lucene/dspace-api/src/main/java/org/dspace/app/util/OpenSearchServiceImpl.java#L87) and uses `dspace.url`, are you sure you've set that correctly on your box (and invalidated the cocoon cache just in case)?
* I'm probably missing something that is being changed by this PR, but my scvccontext is `open-search/discover` (sans trailing slash but note open-search at the beginning). This is backed up by the sitemap [here](https://github.com/DSpace/DSpace/pull/1235/files#diff-058215c1e5b20c63dcfe10d59a0d2f96R548); the opensearch generator is called for the open-search URL pattern.

(I haven't actually tested the PR at all, my comments are based on the changes I made to one of my repos that's on a modified 5.1 codebase)

**Comment by tdonohue** [comment]
@hardyoyo : The point of this PR is to remove Lucene. Hence, it doesn't touch the index-db-browse stuff (which should be done in a different PR).  

If this PR doesn't fully fix DS-2268, then we can create a new PR to do so. I'd suggest we avoid muddying the waters here.

The goal of this PR is to remove everything related to Lucene. If that works and it doesn't introduce any issues, then I'd recommend we merge this PR and then add additional PRs (as needed) to fix DS-2268 and/or remove 'index-db-browse'. Otherwise, the scope of this PR starts to stretch.

**Comment by mwoodiupui** [comment]
I still see Lucene JARs in the installed instance.  Shouldn't they be gone?  'mvn dependency:analyze' mentions them only as "unused declared dependencies".  Maybe a couple of POMs (dspace-parent, dspace-api, modules/solr) need trimming?

**Comment by mwoodiupui** [comment]
Hm, no, o.d.app.util.IndexVersion requires lucene-core for other reasons.  We might be able to exclude most of all of its dependencies on other Lucene JARs though.

And of course Solr will need scads of Lucene JARs.  But I think that none of the other webapps should need anything that dspace-api doesn't need.

It's probably not enough to delay this PR in any case.

**Comment by mwoodiupui** [comment]
I created a Community, a Collection, and one Item.  Searching seems to work as I would expect.  Browsing by title works as expected.  Browsing by author (either by the letter bar or the first-few-letters box) yields all known values.  Browsing by subject yields one or both known subjects in a pattern that I haven't been able to decipher.

**Comment by mwoodiupui** [comment]
I created a Community, a Collection, and one Item, and tested search and browse.  Now that I understand what behaviors I should expect, they all seem to work well, and I see nothing worrisome in the logs.  +1

**Comment by tdonohue** [comment]
This sits at +2, and is non-controversial and tested by a few Committers (self included).  Merging. 

If any bugs are located relating to Lucene removal, we can log them as separate tickets/PRs.

**Pull request merged**